lib/Backend/arm64/LowerMD.cpp

```diff
@@ -6645,25 +6651,17 @@ LowererMD::LowerInt4MulWithBailOut(
         // Fall through to bailOutLabel
     }
 
-    const auto insertBeforeInstr = checkForNegativeZeroLabel ? checkForNegativeZeroLabel : bailOutLabel;
+    IR::LabelInstr* insertBeforeInstr = checkForNegativeZeroLabel ? checkForNegativeZeroLabel : bailOutLabel;
 
     //check overflow
-    //    CMP_ASR31 r17, dst
-    //    BNE $bailOutLabel
     if(bailOutKind & IR::BailOutOnMulOverflow || bailOutKind == IR::BailOutOnFailedHoistedLoopCountBasedBoundCheck)
     {
-        // (SMULL doesn't set the flags but we don't have 32bit overflow <=> r17-unsigned ? r17==0 : all 33 bits of 64bit result are 1's
-        //      CMP r17, dst, ASR #31 -- check for overflow (== means no overflow)
-        IR::RegOpnd* opndRegScratch = IR::RegOpnd::New(nullptr, SCRATCH_REG, TyMachReg, instr->m_func);
-        insertInstr = IR::Instr::New(Js::OpCode::CMP_ASR31, instr->m_func);
-        insertInstr->SetSrc1(opndRegScratch);
-        insertInstr->SetSrc2(dst);
-        insertBeforeInstr->InsertBefore(insertInstr);
-        LegalizeMD::LegalizeInstr(insertInstr, false);
+        insertInstr = IR::Instr::New(Js::OpCode::CMP_SXTW, nullptr, s3, s3, instr->m_func);
+        instr->InsertBefore(insertInstr);
```

Comment:
It seems to break convention for CMP/TST instructions to use nullptr for the dst in the New instead of creating the instruction and setting the two srcs afterward (though I admit I like your way better).

Reply:
Also, I think you removed the final instance of CMP_ASR31, so we should nuke it from MdOpcodes.h and EncoderMD.cpp. I think CMP_SXTW is the better approach on 64-bit.

Reply:
Since the other case of CMP_ASR31 is removed in your pending commit, I'll wait until it's merged, rebase and remove the opcode.